Adds full connection cycle capability to the Connection package.

zinigor opened the pull request:
#### Changes proposed in this Pull Request:
[x] Adds the register method to the Connection package.
[x] Hooks the register verification method for registration verification.
[x] Adds multisite registration capability.
[x] Makes sure the methods are moved with only the needed dependencies.
[x] Makes sure the example plugin is able to connect using only the dependent packages and minimal boilerplate code.

#### Is this a new feature or does it add/remove features to an existing part of Jetpack?
* This PR aims to make the Connection package usable without the Jetpack plugin itself.

#### Testing instructions:
* Use [the example client plugin](https://github.com/Automattic/client-example) on a clean WordPress site. It's bare bones, it has one administration page where it displays some debugging information, as well as lets you press a button to run the registration process. You should see that the site properly generates secrets, but can't proceed with user registration because handlers are not yet in place.

#### Proposed changelog entry for your changes:
<!-- Please do not leave this empty. If no changelog entry needed, state as such. -->
* N/A

zinigor committed:
Moved the register method to the connection package.

tyxla commented on packages/connection/src/Manager.php:
```diff
@@ -124,7 +137,281 @@ public function verify_signature() {
+				'stats_id'        => $stats_id,
+				'state'           => get_current_user_id(),
+				'site_created'    => $this->get_assumed_site_creation_date(),
+				'jetpack_version' => JETPACK__VERSION,
```
Should we use the constants package for this one?

zinigor replied:
Yes, I'm hoping to merge Michael's #12717 into this PR once it gets merged.

zinigor replied:
Fixed that, thank you!

tyxla commented on packages/connection/src/Manager.php:
```diff
@@ -124,7 +137,281 @@ public function verify_signature() {
+		$args['body'] = $this->apply_activation_source_to_args( $args['body'] );
+
+		// TODO: fix URLs for bad hosts
+		$response = \Jetpack_Client::_wp_remote_request(
```
What about this one? Are we planning to move it to a package?

zinigor replied:
Yes, I'm hoping to merge Michael's #12717 into this PR once it gets merged.

gravityrail commented on jetpack.php:
```diff
@@ -214,7 +214,6 @@ function jetpack_admin_missing_autoloader() { ?>
 // @todo: Abstract out the admin functions, and only include them if is_admin()
 require_once( JETPACK__PLUGIN_DIR . 'class.jetpack.php'               );
 require_once( JETPACK__PLUGIN_DIR . 'class.jetpack-network.php'       );
-require_once( JETPACK__PLUGIN_DIR . 'class.jetpack-client.php'        );
```
This line disappearing is my favourite thing about this PR

zinigor committed:
Added constants and defaults to the API url method.

zinigor committed:
Moved a significant part of XMLRPC controller to the connection package.

zinigor committed:
Deprecated a copied method and moved hooks into the package.

zinigor committed:
Add the XMLRPC verification property to the class.

zinigor committed:
Moved register code to the package, including multisite register.

zinigor committed:
Added connection manager passing to the network manager.

mdawaffe commented:
Example client plugin worked for me, and what I understand about the changes make sense :)

mdawaffe committed:
Subsite registration: send the main site's blog_id as `network_wpcom_blog_id`

mdawaffe commented:
491016a3f344e92d3f25bb28046d4dc52a51db70 moved the `$network_wpcom_blog_id = Jetpack_Options::get_option( 'id' );` to after the `switch_to_blog()`, so we're sending the subsite's wpcom_blog_id (which doesn't exist yet) in the subsite registration request. We should be sending the main site's wpcom_blog_id.

3ecdad75cf619f60dbb8c64294f7e133e7fb046f provides a sort of hacky fix (by unswitching/reswitching the blog).

zinigor committed:
Added the legacy compat file back.

zinigor committed:
Fixed tests.

zinigor committed:
Fixed sync tests.

zinigor committed:
Fixed an incorrect escape call.

zinigor committed:
Removed a deprecated function usage.

mdawaffe commented on packages/connection/src/Manager.php:
```diff
@@ -43,6 +160,313 @@ public function initialize( $methods ) {
+		 * @param Array $post_data request data.
+		 * @param Array $token_data token data.
+		 */
+		return apply_filter(
```
Typo.

```suggestion
		return apply_filters(
```

zinigor replied:
Thank you, I have found that one because of breaking tests too.

mdawaffe commented on packages/connection/legacy/class.jetpack-xmlrpc-server.php:
```diff
@@ -577,7 +648,7 @@ function verify_action( $params ) {
 	 */
-	function login() {
+	public function login() {
 		Jetpack::init()->require_jetpack_authentication();
```
This got refactored.

```suggestion
		Jetpack::connection()->require_jetpack_authentication();
```

mdawaffe commented on packages/connection/src/Manager.php:
```diff
@@ -43,6 +160,313 @@ public function initialize( $methods ) {
+	 * @return WP_User|Mixed authenticated user or error.
+	 */
+	public function authenticate_jetpack( $user, $username, $password ) {
+		if ( is_a( $user, 'WP_User' ) ) {
```
This code is namespaced now, so we need to be explicit here.
```suggestion
		if ( is_a( $user, '\WP_User' ) ) {
```

mdawaffe commented on packages/connection/src/Manager.php:
```diff
@@ -43,6 +160,313 @@ public function initialize( $methods ) {
+
+		nocache_headers();
+
+		return new WP_User( $token_details['user_id'] );
```
…and here (namespace).
```suggestion
		return new \WP_User( $token_details['user_id'] );
```

zinigor committed:
Fixed REST signature checking.

zinigor committed:
Update packages/connection/legacy/class.jetpack-xmlrpc-server.php

Co-Authored-By: Michael D Adams <michael.d.adams@gmail.com>

zinigor committed:
Update packages/connection/src/Manager.php

Co-Authored-By: Michael D Adams <michael.d.adams@gmail.com>

zinigor committed:
Update packages/connection/src/Manager.php

Co-Authored-By: Michael D Adams <michael.d.adams@gmail.com>

zinigor committed:
Added a default blog token value.

mdawaffe committed:
Signature Verification: Modify filtered value

We don't have enough data to (easily) recreate the value passed to the
`jetpack_signature_check_token` filter. Instead, modify the existing value
with the new data.

mdawaffe committed:
Sync Tests: Reset Connection Manager's state

Make sure the Connection_Manager's `->raw_post_data` state is reset after
tests which modify `$GLOBALS['HTTP_RAW_POST_DATA']`.

zinigor committed:
Fixed broken tests due to refactor.

tyxla commented on packages/connection/src/Manager.php:
```diff
@@ -28,6 +29,155 @@ class Manager implements Manager_Interface {
+	/**
+	 * Creates an instance of the connection manager.
+	 */
+	public function __construct() {
```
Do we need a constructor if it doesn't do anything? Seems like we could remove it.

tyxla commented on packages/connection/src/Manager.php:
```diff
@@ -28,6 +29,155 @@ class Manager implements Manager_Interface {
+		if (
+			! $is_jetpack_xmlrpc_request
+			&& is_admin()
+			&& isset( $_POST['action'] ) // phpcs:ignore WordPress.Security.NonceVerification
```
Definitely not a requirement, but I like adding `@todo` blocks in the docblocks of the methods when there is something that needs refactoring. It would be nice to implement a better nonce verification here at some point.

tyxla replied:
Done in 0dd9046c77297db60bb895f86b190e148c5f9dcf

tyxla commented on packages/connection/src/Manager.php:
```diff
@@ -28,6 +29,155 @@ class Manager implements Manager_Interface {
+		) {
+			$this->require_jetpack_authentication();
+			$this->add_remote_request_handlers();
+		} else {
```
Could we use `return` at the end of the `if()` block and that way prevent from the need to have everything in an `else` block?

tyxla replied:
Done in b5bc77afac0b66a68e75e2fbb22b2dba113c2fbe

tyxla commented on packages/connection/src/Manager.php:
```diff
@@ -28,6 +29,155 @@ class Manager implements Manager_Interface {
+			return false;
+		}
+		// Display errors can cause the XML to be not well formed.
+		@ini_set( 'display_errors', false ); // phpcs:ignore
```
😅

tyxla commented on packages/connection/src/Manager.php:
```diff
@@ -28,6 +29,155 @@ class Manager implements Manager_Interface {
+		if ( $xmlrpc_server ) {
+			$this->xmlrpc_server = $xmlrpc_server;
+		} else {
+			require_once JETPACK__PLUGIN_DIR . 'class.jetpack-xmlrpc-server.php';
```
Do we have a plan to remove this one? Seems to be still coupling us to the plugin.

tyxla replied:
FWIW I just hit a fatal here when trying the client example plugin:

```
[05-Aug-2019 10:50:04 UTC] PHP Warning:  Use of undefined constant JETPACK__PLUGIN_DIR - assumed 'JETPACK__PLUGIN_DIR' (this will throw an Error in a future version of PHP) in //wp-content/plugins/client-example/vendor/automattic/jetpack-connection/src/Manager.php on line 131
[05-Aug-2019 10:50:04 UTC] PHP Warning:  require_once(JETPACK__PLUGIN_DIRclass.jetpack-xmlrpc-server.php): failed to open stream: No such file or directory in //wp-content/plugins/client-example/vendor/automattic/jetpack-connection/src/Manager.php on line 131
[05-Aug-2019 10:50:04 UTC] PHP Fatal error:  require_once(): Failed opening required 'JETPACK__PLUGIN_DIRclass.jetpack-xmlrpc-server.php' (include_path='.:/opt/sp/php7.2/lib/php') in //wp-content/plugins/client-example/vendor/automattic/jetpack-connection/src/Manager.php on line 131
```

Maybe we should try to provide an `$xmlrpc_server` object before running this one? Also, we could totally use the `Constants` package for handling the constant, and I doubt we'll want to include a file from Jetpack, so we'll indeed need to make the xmlrpc server class independent.

tyxla replied:
I was able to make this work by commenting this line, as we'll load the class as part of the connection package now:

```suggestion
			//require_once JETPACK__PLUGIN_DIR . 'class.jetpack-xmlrpc-server.php';
```

tyxla replied:
Committed that separately, marking as resolved.

tyxla commented on packages/connection/src/Manager.php:
```diff
@@ -28,6 +29,155 @@ class Manager implements Manager_Interface {
+				// The jetpack Provision method is available for blog-token-signed requests.
+				add_filter( 'xmlrpc_methods', array( $this->xmlrpc_server, 'provision_xmlrpc_methods' ) );
+			} else {
+				new XMLRPC_Connector( $this );
```
Silly question perhaps, but do we need to do anything with the connector object references? Seems like we don't use them anywhere and are only interested in instantiating them.

tyxla replied:
I think we're good to keep this as-is.

tyxla commented on packages/connection/src/Manager.php:
```diff
@@ -43,6 +193,313 @@ public function initialize( $methods ) {
+		// Some browser-embedded clients send cookies. We don't want them.
+		$_COOKIE = array();
+
+		// A bug in PHP < 5.2.2 makes $HTTP_RAW_POST_DATA not set by default,
```
Given our current PHP min version support, I wonder if we could finally get rid of that hack.

tyxla replied:
Removed 👍

tyxla commented on packages/connection/src/Manager.php:
```diff
@@ -43,6 +193,313 @@ public function initialize( $methods ) {
+		 * @param string $class The name of the XML-RPC server class.
+		 */
+		$wp_xmlrpc_server_class = apply_filters( 'wp_xmlrpc_server_class', 'wp_xmlrpc_server' );
+		$wp_xmlrpc_server       = new $wp_xmlrpc_server_class();
```
This looks a bit fragile. Could we have a whitelist of classes we allow here? I'm concerned anyone could hook a malicious class in the `wp_xmlrpc_server_class` filter that could do bad bad things.

tyxla replied:
Added a `@todo` to address this in 5a4a72b6ac6e2610fb3f47041b762ae0e118ea60. It was like that before, so no rush to fix it now.

tyxla commented on packages/connection/src/Manager.php:
```diff
@@ -43,6 +193,313 @@ public function initialize( $methods ) {
+	 * @return WP_User|Mixed authenticated user or error.
+	 */
+	public function authenticate_jetpack( $user, $username, $password ) {
+		if ( is_a( $user, '\WP_User' ) ) {
```
We need a double backslash here because of how they are parsed within single quotes.

```suggestion
		if ( is_a( $user, '\\WP_User' ) ) {
```

tyxla commented on packages/connection/src/Manager.php:
```diff
@@ -43,6 +193,313 @@ public function initialize( $methods ) {
+				 * - invalid_nonce
+				 * - signature_mismatch
+				 *
+				 * @since 7.5.0
```
Another case that we should correct to the actual current version we're working towards?

tyxla replied:
Nope, this was indeed introduced in Jetpack 7.5.0.

tyxla commented on packages/connection/src/Manager.php:
```diff
@@ -43,6 +193,313 @@ public function initialize( $methods ) {
+	 * This function has side effects and should not be used. Instead,
+	 * use the memoized version `->verify_xml_rpc_signature()`.
+	 *
+	 * @internal
```
Could also specify `@access private` for all those internal methods.

tyxla replied:
Not necessary for now.

tyxla commented on packages/connection/src/Manager.php:
```diff
@@ -145,15 +602,412 @@ public function verify_signature() {
-		return 0;
+	public function register( $api_endpoint = 'register' ) {
+
+		add_action( 'pre_update_jetpack_option_register', array( '\Jetpack_Options', 'delete_option' ) );
```
```suggestion
		add_action( 'pre_update_jetpack_option_register', array( '\\Jetpack_Options', 'delete_option' ) );
```

tyxla commented on packages/connection/src/Manager.php:
```diff
@@ -145,15 +602,412 @@ public function verify_signature() {
-	public function register() {
-		return 0;
+	public function register( $api_endpoint = 'register' ) {
+
```
```suggestion
```

tyxla commented on packages/connection/src/Manager.php:
```diff
@@ -145,15 +602,412 @@ public function verify_signature() {
+		} elseif ( ! $registration_details ) {
+			return new \WP_Error(
+				'unknown_error',
+				'',
```
Maybe we could specify where/when the error occurred? Like `An error occurred during attempt to validate remote registration`.

tyxla replied:
Added an error message here.

tyxla commented on packages/connection/src/Manager.php:
```diff
@@ -145,15 +602,412 @@ public function verify_signature() {
+		if ( empty( $registration_details->jetpack_secret ) || ! is_string( $registration_details->jetpack_secret ) ) {
+			return new \WP_Error(
+				'jetpack_secret',
+				'',
```
Maybe we could specify more information about the error? Like `A Jetpack secret is required for validating registration` or something.

tyxla replied:
Added an error message here.

tyxla commented on packages/connection/src/Manager.php:
```diff
@@ -145,15 +602,412 @@ public function verify_signature() {
+
+		$sql .= ' ORDER BY `option_id` LIMIT 100';
+
+		$sql = $wpdb->prepare( $sql, $sql_args ); // phpcs:ignore
```
Maybe we can be a bit more specific about the rules we want to ignore?

tyxla replied:
Done.

tyxla commented on packages/connection/src/Manager.php:
```diff
@@ -660,4 +1514,117 @@ public function get_access_token( $user_id = false, $token_key = false, $suppres
+		$user_data = $this->get_connected_user_data();
+		if ( is_array( $user_data ) ) {
+			$options['jetpack_user_id']         = array(
+				'desc'     => __( 'The WP.com user ID of the connected user' ), // phpcs:ignore WordPress.WP.I18n.MissingArgDomain
```
Seems like we could ignore that i18n rule for the entire block of code, instead of doing it several times.

tyxla replied:
I think that's fine; I've also added a `@todo` block to audit whether those even make sense - it seems to me they should be using a textdomain, too. Auditing is not required now, and can be done separately.

tyxla commented on packages/connection/tests/php/test_Manager.php:
```diff
@@ -49,6 +52,32 @@ public function test_is_active_when_not_connected() {
 		$this->assertFalse( $this->manager->is_active() );
 	}
 
+	public function test_api_url_defaults() {
```
Would be great to add some more tests (and remove tests that no longer make sense to be in the default Jetpack suite), but that's definitely not a requirement for this PR; we can do it in another PR.

tyxla commented on tests/php/bootstrap.php:
```diff
@@ -118,6 +118,8 @@ function in_running_uninstall_group() {
 	return is_array( $argv ) && in_array( '--group=uninstall', $argv );
 }
 
+require dirname( __FILE__ ) . '/../../vendor/autoload.php';
```
We should be able to use `__DIR__` here directly, since we no longer care about older PHP versions that didn't support it.

tyxla replied:
Done in 34f7c850c5d03b48c82a4583f7e3daa6e837770c

tyxla commented on tests/php/general/test_class.jetpack-network.php:
```diff
@@ -89,15 +89,6 @@ public function test_get_url_returns_null_for_underspecified_subsitedisconnect()
-	/**
-	 * @since 2.8
-	 **/
-	public function test_set_auto_activated_modules_returns_array() {
```
Should we port this test somewhere, or is there a good reason we're removing it?

tyxla replied:
No need to port it, we've removed the method.

tyxla commented on tests/php/general/test_class.jetpack-xmlrpc-server.php:
```diff
@@ -2,8 +2,6 @@
 
 use Automattic\Jetpack\Sync\Sender;
 
-require_once dirname( __FILE__ ) . '/../../../class.jetpack-xmlrpc-server.php';
```
👍

tyxla commented on tests/php/sync/test_class.jetpack-sync-callables.php:
```diff
@@ -1001,6 +1002,7 @@ function mock_authenticated_xml_rpc_cleanup( $user_id ) {
 		unset( $_SERVER['REQUEST_METHOD'] );
 		$jetpack = Jetpack::init();
 		$jetpack->reset_saved_auth_state();
```
Should we use the `reset_saved_auth_state` from the connection manager here too?

tyxla replied:
No, this will also cleanup the rest auth status, so let's leave it as-is for now.

tyxla reviewed (commented):
Wow, this is amazing work, thank you for pulling it off! 

I'm yet to test it properly (will do it now), but I'm sending some feedback in the meantime. It's mostly minor stuff, really.

tyxla commented on packages/connection/src/Manager.php:
```diff
@@ -28,6 +29,155 @@ class Manager implements Manager_Interface {
+			$this->xmlrpc_server = $xmlrpc_server;
+		} else {
+			require_once JETPACK__PLUGIN_DIR . 'class.jetpack-xmlrpc-server.php';
+			$this->xmlrpc_server = new Jetpack_XMLRPC_Server();
```
I was able to make this work by telling Composer to look for it in the global namespace:
```suggestion
			$this->xmlrpc_server = new \Jetpack_XMLRPC_Server();
```

tyxla reviewed (commented):
I made my way around this first error, see the comments:

tyxla commented on packages/connection/src/Manager.php:
```diff
@@ -660,4 +1514,117 @@ public function get_access_token( $user_id = false, $token_key = false, $suppres
+		$options['jetpack_version'] = array(
+			'desc'     => __( 'Jetpack Plugin Version' ), // phpcs:ignore WordPress.WP.I18n.MissingArgDomain
+			'readonly' => true,
+			'value'    => JETPACK__VERSION,
```
Got another warning here:

```
[05-Aug-2019 11:02:35 UTC] PHP Warning:  Use of undefined constant JETPACK__VERSION - assumed 'JETPACK__VERSION' (this will throw an Error in a future version of PHP) in /wp-content/plugins/client-example/vendor/automattic/jetpack-connection/src/Manager.php on line 1613
```

Should be resolved if we use the Constants package:

```suggestion
			'value'    => Constants::get_constant( 'JETPACK__VERSION' ),
```

lezama commented on class.jetpack-network.php:
```diff
@@ -407,164 +421,92 @@ public function do_subsiteregister( $site_id = null ) {
 			return;
 		}
 
-		$jp = Jetpack::init();
```
it's probably already initialised at this point I guess?

tyxla replied:
Yes 👍

tyxla commented:
Jumping in to address feedback here and push it forward.

tyxla committed:
Sort the use clauses

tyxla committed:
Use proper version in docs

tyxla commented on class.jetpack-network.php:
```diff
@@ -74,35 +89,19 @@ private function __construct() {
 			 * This is a hacky way because xmlrpc is not available on wp_initialize_site
 			 */
-			if ( $this->get_option( 'auto-connect' ) == 1 ) {
+			if ( 1 === $this->get_option( 'auto-connect' ) ) {
```
I've verified and it's not possible; this is a serialized array, so it'll keep its integer type.

tyxla committed:
Merge comments

tyxla commented on class.jetpack-network.php:
```diff
@@ -407,164 +421,92 @@ public function do_subsiteregister( $site_id = null ) {
+	 *
+	 * @param Object $token the received token.
+	 */
+	public function filter_register_user_token( $token ) {
```
Actually, the hook we hook this to is an action, so returning is not needed. I'll just change the `add_filter` to `add_action`.

tyxla committed:
Hook filter_register_user_token properly

tyxla commented on class.jetpack-network.php:
```diff
@@ -407,164 +421,92 @@ public function do_subsiteregister( $site_id = null ) {
-		 * This is in contrast to regular standalone sites, where we use the helper
-		 * `Jetpack::get_assumed_site_creation_date()` to assume the site's creation date.
-		 */
+		$is_master_user = ! Jetpack::is_active();
```
Going to keep it as-is for now, so we keep changes as minimal as possible.

tyxla committed:
Rename network sites table var

tyxla committed:
Use esc_html__()

tyxla committed:
Fix code in comment

tyxla committed:
Add a deprecation todo comment

tyxla commented on class.jetpack.php:
```diff
@@ -721,70 +701,26 @@ private function __construct() {
+	 * @param Boolean               $is_signed whether the signature check has been successful.
+	 * @param Jetpack_XMLRPC_Server $xmlrpc_server (optional) an instance of the server to use instead of instantiating a new one.
+	 */
+	public function setup_xmlrpc_handlers(
```
Let's leave that for another time - deprecating it will require updating tests, and I'd like to avoid making this PR larger. Adding a `@todo` for that.

tyxla committed:
Align comment descriptions vertically

tyxla committed:
Update variable to follow the naming conventions

tyxla commented on class.jetpack.php:
```diff
@@ -5224,217 +5020,8 @@ public static function load_xml_rpc_client() {
 	 */
 	public function reset_saved_auth_state() {
-		$this->xmlrpc_verification = null;
 		$this->rest_authentication_status = null;
```
Will keep it as-is for now, as I'm not sure it should be changed. Seems like it's intentionally left here.

lezama commented on class.jetpack.php:
```diff
@@ -5523,143 +5110,6 @@ public function wp_rest_authentication_errors( $value ) {
-		return $wp_xmlrpc_server->wp_getOptions( $args );
-	}
-
-	function xmlrpc_options( $options ) {
```
hmm, let's do it just in case, but probably no one is using them 🤷‍♂

tyxla replied:
Done in 77069b4f2c317cd15dd48e960640b231b6625f35

lezama commented on packages/connection/legacy/class.jetpack-client.php:
```diff
@@ -0,0 +1,5 @@
+<?php
```
let's delete this one :+1:

lezama commented on packages/connection/legacy/class.jetpack-xmlrpc-server.php:
```diff
@@ -1,6 +1,7 @@
-<?php
+<?php // phpcs:ignore
```
👍

lezama committed:
Update packages/connection/legacy/class.jetpack-xmlrpc-server.php

Co-Authored-By: Marin Atanasov <8436925+tyxla@users.noreply.github.com>

tyxla committed:
Deprecate set_auto_activated_modules

tyxla committed:
Deprecate methods instead of removing them completely

tyxla committed:
Cleanup unnecessary method

tyxla committed:
Remove unnecessary compat class file

tyxla committed:
Add a file doc comment

tyxla committed:
Add a compat file for class.jetpack-xmlrpc-server

tyxla committed:
Align some docblocks

tyxla committed:
Fully qualified names for classes defined in global namespace

tyxla committed:
Remove empty constructor

tyxla committed:
Add some @todos for potential improvements

tyxla committed:
Prevent from a nested if

tyxla committed:
We no longer need to require the xmlrpc server class manually

tyxla committed:
Instantiate class from the global namespace

tyxla committed:
🔪 Remove PHP 5.2.2 hack (see https://bugs.php.net/bug.php?id=41293)

tyxla committed:
Double backslash to properly reference the class from the global namespace

tyxla committed:
Add a todo to tighten wp_xmlrpc_server_class

tyxla committed:
Remove an extra empty line

tyxla committed:
Use double backslash to specify global namespace properly

tyxla committed:
[not verified] More specific error messages during registration

tyxla committed:
[not verified] Be more specific with PHPCS ignores

tyxla committed:
Use Constants package instead of directly referencing a Jetpack constant

tyxla committed:
Statically access verify_xml_rpc_signature to make sure connection is initialized

tyxla committed:
Always access verify_xml_rpc_signature from the connection package

tyxla committed:
No need to init when calling ::connection()

tyxla committed:
Make use of __DIR__

tyxla committed:
Add Jetpack_XMLRPC_Server to the autoloader ignore list

tyxla committed:
Use api_url from the connection package in Jetpack_Client_Server

tyxla commented:
I think this is ready for another review!

More info on how to test it in p1HpG7-7lI-p2 and its comments.

lezama reviewed (dismissed):
Things look to still connect and work, let's merge early and keep iterating and testing around this through the entire current cycle.

Great work @zinigor and @tyxla!

lezama committed:
don't require class.jetpack-xmlrpc-server.php

lezama reviewed (approved):
re approving after my last commit :)
